scripts/generate_rust_analyzer.py

```diff
@@ -0,0 +1,175 @@
+            cfg=[]
+        )
+    )
+
+    # compiler_builtins
+    ret.append(
+        generate_crate(
+            name="compiler_builtins",
+            module=srctree + "/rust/compiler_builtins.rs",
+            deps=[],
+            cfg=[]
+        )
+    )
+
+    # alloc
+    ret.append(
+        generate_crate(
+            name="alloc",
+            module=lib_src + "/alloc/src/lib.rs",
+            member=False,
+            deps=generate_deps(ret, ["core", "compiler_builtins"]),
+            cfg=[]
+        )
+    )
+
+    # module
+    module = generate_crate(
+        name="module",
+        module=srctree + "/rust/module.rs",
+        deps=[],
+        cfg=[],
+    )
+    module["proc_macro_dylib_path"] = "rust/libmodule.so"
+    ret.append(
+        module
+    )
+
+    # kernel
+    kernel = generate_crate(
+        name="kernel",
+        module=srctree + "/rust/kernel/lib.rs",
+        deps=generate_deps(ret, ["core", "alloc", "module"]),
+    )
+    kernel["env"]["RUST_BINDINGS_FILE"] = bindgen_file
+    kernel["source"] = {
+        "include_dirs": [ srctree + "/rust/kernel", objtree + "/rust" ],
+        "exclude_dirs": []
+    }
+    ret.append(kernel)
+
+    return ret
+
+def does_file_contains(filepath, search):
+    with open (filepath) as f:
+        for line in f:
+            if search in line:
+                return True
+    return False
+
+def check_and_generate(filepath, source=crate_list):
+    file = path.basename(filepath)
+    makefile = path.dirname(filepath) + "/Makefile"
+    name = file.replace('.rs', '')
+    objname = name + ".o"
+    print("checking " + objname, file=sys.stderr)
+    if does_file_contains(makefile, objname):
+        print("building crate " + name, file=sys.stderr)
+        return generate_crate(
+            name=name,
+            module=filepath,
+            deps=get_default_deps(source=source)
+        )
+
+
+def find_rs(tree):
+    ret = []
+    for root, dirs, files in os.walk(tree):
+        for file in files:
+            if file.endswith(".rs"):
+                ret.append(path.join(root, file))
+    return ret
+
+def generate_samples(srctree=srctree, source=crate_list):
+    ret = []
+    for x in find_rs(srctree + "/samples/rust"):
+        sample = check_and_generate(x, source=source)
+        if sample:
+            ret.append(sample)
+    return ret
+
+def generate_drivers(srctree=srctree, source=crate_list):
+    ret = []
+    for x in find_rs(srctree + "/drivers"):
+        driver = check_and_generate(x, source=source)
+        if driver:
+            ret.append(driver)
+    return ret
+
+
+
+if __name__ == "__main__":
+    crate_list = crate_list + generate_kernel()
+    crate_list = crate_list + generate_samples(srctree=srctree, source=crate_list)
+    crate_list = crate_list + generate_drivers(srctree=srctree, source=crate_list)
+
+    rust_project = {
+        "crates": crate_list,
+        "sysroot_src": lib_src
+    }
+    print(json.dumps(rust_project))
```

Comment:
Pretty printing is trivial now that we use Python's `json` and is better to debug the file and for human readers etc.:

```suggestion
    print(json.dumps(rust_project, sort_keys=True, indent=4))
```

Reply:
Is sort_keys safe? (guess it is, but would be fatal if it reorders the crates array)

Reply:
Yeah, it is for dictionary keys, not for the ordered lists.

Reply:
By the way, if you want to avoid the string intermediate, you can also write directly to `stdout`:

```python
json.dump(rust_project, sys.stdout, sort_keys=True, indent=4)
```